R/cube-comparisons.R

```diff
@@ -0,0 +1,306 @@
+#'
+#' *Warning* since there is more than one comparison being made against each
+#' baseline the z-scores, and especially the p-values derived from these
+#' z-scores should be interpreted with caution. Using standard p-value cutoffs
+#' will result in anti-conservative interpretations because of the \href{https://en.wikipedia.org/wiki/Multiple_comparisons_problem}{multiple
+#' comparisons problem}.
+#' Adjustments to p-value cut offs (e.g. \href{https://en.wikipedia.org/wiki/Bonferroni_correction}{Bonferonni correction}) should be used when interpreting z-scores from the
+#' `compare[Rows|Cols|Dims]Pairwise()` family of functions.
+#'
+#' @param cube a CrunchCube to calculate the comparison on
+#' @param dim which dimension is being compared (`rows` or `cols`, only valid
+#'   for `compareDims()`)
+#' @param baseline a character, the name of a column or row to compare
+#'   against each of the others in turn
+#' @param value Value of the test to return. Defaults to `statistic`; use `p.value`
+#' for the p-value of the test of the pair of columns/rows.
+#' @param ... arguments passed from `compareRowsPairwise()` or
+#'   `compareColsPairwise()` to `compareDimsPairwise()` (e.g., `baseline`)
+#'
+#' @return a vector of representing the baseline column’s test statistic
+#' with respect to each of the others.
+#'
+#' @name dimension-comparison-pairwise
+NULL
+
+#' @rdname dimension-comparison-pairwise
+#' @export
+compareColsPairwise <- function(cube, ...) {
+    compareDimsPairwise(cube = cube, dim = "cols", ...)
+}
+
+#' @rdname dimension-comparison-pairwise
+#' @export
+compareRowsPairwise <- function(cube, ...) {
+    compareDimsPairwise(cube = cube, dim = "rows", ...)
+}
+
+#' @rdname dimension-comparison-pairwise
+#' @export
+compareDimsPairwise <- function(cube, dim = c("cols", "rows"), baseline, value=c("statistic", "p.value")) {
+    dim <- match.arg(dim)
+    
+    # grab the names of extents for each dimensions
+    # TODO: we shouldn't need to do this if we can just pass baseline/x into the
+    # subsetting function when we can subset by name _or_ id natively
+    if (dim == "cols") {
+        names <- colnames(as.array(cube))
+    } else if (dim == "rows") {
+        names <- rownames(as.array(cube))
+    }
+    
+    to_compare <- names[!(names %in% baseline)]
+    
+    out <- vapply(
+        names, function(one_extent) {
+            # generate the 2xm or nx2 table for testing
+            return(compareDims(cube, baseline = baseline, x = one_extent, dim = dim, value=value))
+        }
+        , numeric(1)
+    )
+    
+    out <- simplify2array(out)
+    
+    # if the dimension is rows, the array that vapply returns needs to be
+    # transposed
+    if (dim == "rows") {
+        out <- t(out)
+    }
+    
+    names(out) <- names
+    
+    return(out)
+}
+
+#' Matrix of Chi-Squared Statistics for all rows or columns
+#' 
+#' Generate a matrix of pairwise comparisons of rows or columns, each against 
+#' the others.
+#'
```

Comment:
To add:

The null hypothesis is that for each pair of rows (or columns) those two specific rows (or columns) in the pair are equal to each other. The test statistic matrix that is returned when requested is a measure of closeness between the pair of rows (or columns) given by their names. The p-value matrix that is returned are similarly the probabilities of finding the observed or more extreme results while the null hypothesis is true for each pair of rows (or columns).